### ISSUE TITLE
Odd cluster center behavior in C-Means 

### ISSUE THREAD

**Issue by JDWarner** [scikit-fuzzy/scikit-fuzzy#94]
[From the mailing list](https://groups.google.com/forum/#!topic/scikit-fuzzy/YHvFKCXb4BQ), reported by Aiken Oliver:

I'm looking as using fuzzy cmeans in some projects, and particularly with data with a large number of clusters. As a first pass, I'm generating a number of blobs using make_blobs out of sklearn.datasets: 

``` python
X, y = make_blobs(n_samples=5000,
                  n_features=8,
                  centers=25,
                  cluster_std=2,
                  center_box=(-100.0, 100.0),
                  shuffle=True)

centers, clusters, _, _, _, _, _ = fuzz.cmeans(X.T, 25, 2, 0.0005, 2000)
```

Passing the resulting blobs to cmeans for clustering results appears to work when 15 or fewer centers and 8 features. Increasing the number of features or centers by more than a few causes many or all of the cmeans centers to all converge to overlapping points. This seems to occur whether the increase is to the number of features or centers -- either will cause this convergence. I didn't see a discussion of this in the papers on fcm I've looked through, but was wondering if this was a behavior of the algorithm that is expected. Perhaps it can be overcome by careful selection of the initial partition matrix? But I've not been able to do so this far...

**Comment by JDWarner** [comment]
This is interesting. The cluster centers should automatically repulse each other. However, it may be that if the blob density is sufficiently high such that it's likely more than one cluster center will fall on the same blob, they will partition/split the blob instead of finding new blobs. 

If true, this should be an issue largely confined to toy examples, as blobs in a higher dimensional feature space should make blobs quite sparse. If this is what's going on, the effect may be able to be mitigated by tweaking the cluster center initialization, such as initializing centers with a pseudo-random space filling approach.

**Comment by sildar** [comment]
Hi!

I don't think this is confined to toy examples. I'm clustering texts, and we typically have a lot of features and lots of clusters. In my use case, I have ~6.000 texts and ~2.000 clusters. Texts have in the tens of thousands of features.

When running the code, _every_ center point converges, so I end up with a 0.0005 probability for each center, for each text.

Things I tried so far:
- Checked my code, the toy example on the mailing list works fine and I followed exaclty the same process for the 6.000 texts / 2.000 clusters / 25.000 features.
- Tried to remove 0 values in the features. It does not solve the problem.
- Tried to reduce the dimensionality (with PCA). Works when reducing to 5-10 features but it would be silly to reduce the dimensionality that much. At 10 features there is already a convergence of the centers.
- Tried a different distance measure. I tried both Euclidean distance and cosine distance. No more success with cosine.
- Checked the maximum error and maximum iterations parameters. They are correct, and the algorithm does converge quickly anyway (as would be expected with 2.000 clusters, and as expected with the bug as centers all converge towards the same point).

I'll try to initialize the centers and see what it gives.
I doubt this behaviour is "by design", but it would be useful to state in the documentation that the code doesn't work for more than a few features and a few clusters.
Unfortunately, it's quite long to run with that quantity of data, so I don't know if I'll be able to dig into it a bit more. I'll try to initialize the centers to some of my data points. I expect it should solve the problem, I'll let you know.

**Comment by sildar** [comment]
As I suspected, initializing the centers to some of my data points "kinda" solves the problem. I took the first 2.000 texts as initial centers. These end up with a very high probability to belong to their cluster (expected), but very very few other texts are added to these clusters. Out of the 4.000 remaining texts, 20 only have a probability > 0.01 to belong to one of the predefined clusters. Maybe this is expected, but it seems strange to have a constant probability for each cluster center whatever the text may be. They basically all have a prob of 1/2.000, sometimes a little bit more (e.g. 0.004 instead of 0.0005).
I'll try and lower the number of expected clusters to something around 200, I want to check if things improve with a better text/clusters ratio.
